File: backend/app/api/routes/transactions.py
"""Consulta i edicio dels moviments d'un espai."""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import and_, func, or_, select
from sqlalchemy.orm import Session

from app.core.time import utcnow
from app.deps import CurrentUser, DbSession, EditableWorkspace, Workspace
from app.models import Category, Ledger, LlmSuggestion, Merchant, Transaction
from app.models.enums import CategorySource, TransactionStatus
from app.schemas.common import Message, Page
from app.schemas.transaction import (
    BulkCategorize,
    ReviewItem,
    TransactionOut,
    TransactionUpdate,
)
from app.services.classification import build_learned_rule, remember_merchant_choice

router = APIRouter(prefix="/transactions", tags=["moviments"])


def to_out(transaction: Transaction) -> TransactionOut:
    data = TransactionOut.model_validate(transaction)
    data.category_name = transaction.category.full_name if transaction.category else None
    if transaction.is_masked:
        data.description = transaction.display_description or ""
        data.normalized_description = ""
        data.counterparty = ""
        data.merchant_name = None
        data.is_masked = True
    else:
        data.merchant_name = transaction.merchant.display_name if transaction.merchant else None
        data.is_masked = False
    return data


def search_clause(pattern: str):
    """Cerca el text visible. Un moviment enmascarat no es troba pel concepte del banc."""
    masked = Transaction.display_description.isnot(None)
    return or_(
        and_(
            masked,
            or_(
                Transaction.display_description.ilike(pattern),
                Transaction.notes.ilike(pattern),
            ),
        ),
        and_(
            Transaction.display_description.is_(None),
            or_(
                Transaction.description.ilike(pattern),
                Transaction.normalized_description.ilike(pattern),
                Transaction.counterparty.ilike(pattern),
                Transaction.notes.ilike(pattern),
            ),
        ),
    )


def _apply_filters(
    query,
    *,
    ledger_id: int,
    account_id: int | None,
    date_from: date | None,
    date_to: date | None,
    category_ids: list[int] | None,
    merchant_id: int | None,
    search: str | None,
    min_amount: Decimal | None,
    max_amount: Decimal | None,
    only_review: bool,
    only_uncategorized: bool,
    transaction_status: TransactionStatus | None,
    include_transfers: bool,
):
    query = query.where(Transaction.ledger_id == ledger_id)
    if account_id is not None:
        query = query.where(Transaction.account_id == account_id)
    if date_from is not None:
        query = query.where(Transaction.booking_date >= date_from)
    if date_to is not None:
        query = query.where(Transaction.booking_date <= date_to)
    if category_ids:
        query = query.where(Transaction.category_id.in_(category_ids))
    if merchant_id is not None:
        query = query.where(Transaction.merchant_id == merchant_id)
    if search:
        query = query.where(search_clause(f"%{search.strip()}%"))
    if min_amount is not None:
        query = query.where(Transaction.amount >= min_amount)
    if max_amount is not None:
        query = query.where(Transaction.amount <= max_amount)
    if only_review:
        query = query.where(Transaction.needs_review.is_(True))
    if only_uncategorized:
        query = query.where(Transaction.category_id.is_(None))
    if transaction_status is not None:
        query = query.where(Transaction.status == transaction_status)
    if not include_transfers:
        query = query.where(Transaction.transfer_group_id.is_(None))
    return query


@router.get("", response_model=Page[TransactionOut])
def list_transactions(
    db: DbSession,
    workspace: Workspace,
    account_id: int | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    category_ids: list[int] | None = Query(default=None),
    merchant_id: int | None = None,
    search: str | None = None,
    min_amount: Decimal | None = None,
    max_amount: Decimal | None = None,
    only_review: bool = False,
    only_uncategorized: bool = False,
    transaction_status: TransactionStatus | None = None,
    include_transfers: bool = True,
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
):
    filters = dict(
        ledger_id=workspace.id,
        account_id=account_id,
        date_from=date_from,
        date_to=date_to,
        category_ids=category_ids,
        merchant_id=merchant_id,
        search=search,
        min_amount=min_amount,
        max_amount=max_amount,
        only_review=only_review,
        only_uncategorized=only_uncategorized,
        transaction_status=transaction_status,
        include_transfers=include_transfers,
    )
    total = db.scalar(_apply_filters(select(func.count(Transaction.id)), **filters))
    rows = db.scalars(
        _apply_filters(select(Transaction), **filters)
        .order_by(Transaction.booking_date.desc(), Transaction.id.desc())
        .limit(limit)
        .offset(offset)
    ).all()
    return Page[TransactionOut](
        items=[to_out(item) for item in rows],
        total=int(total or 0),
        limit=limit,
        offset=offset,
    )


@router.get("/review", response_model=Page[ReviewItem])
def review_queue(
    db: DbSession,
    workspace: Workspace,
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
):
    """Cua de moviments pendents de revisar, amb el suggeriment del model."""
    condition = (Transaction.ledger_id == workspace.id) & (Transaction.needs_review.is_(True))
    total = db.scalar(select(func.count(Transaction.id)).where(condition))
    rows = db.scalars(
        select(Transaction)
        .where(condition)
        .order_by(Transaction.booking_date.desc(), Transaction.id.desc())
        .limit(limit)
        .offset(offset)
    ).all()

    suggestions: dict[int, LlmSuggestion] = {}
    merchant_ids = {item.merchant_id for item in rows if item.merchant_id}
    if merchant_ids:
        for suggestion in db.scalars(
            select(LlmSuggestion)
            .where(LlmSuggestion.merchant_id.in_(merchant_ids))
            .order_by(LlmSuggestion.created_at)
        ):
            suggestions[suggestion.merchant_id] = suggestion

    items = []
    for transaction in rows:
        suggestion = suggestions.get(transaction.merchant_id or -1)
        items.append(
            ReviewItem(
                transaction=to_out(transaction),
                suggested_category_id=suggestion.suggested_category_id if suggestion else None,
                suggested_category_name=(
                    suggestion.suggested_category.full_name
                    if suggestion and suggestion.suggested_category
                    else None
                ),
                confidence=suggestion.confidence if suggestion else None,
                rationale=suggestion.rationale if suggestion else "",
            )
        )
    return Page[ReviewItem](items=items, total=int(total or 0), limit=limit, offset=offset)


@router.post("/bulk-categorize", response_model=Message)
def bulk_categorize(
    payload: BulkCategorize, db: DbSession, user: CurrentUser, workspace: EditableWorkspace
):
    """Assigna la mateixa categoria a un conjunt de moviments de l'espai."""
    category = _category_or_400(db, workspace, payload.category_id)

    transactions = list(
        db.scalars(
            select(Transaction).where(
                Transaction.id.in_(payload.transaction_ids),
                Transaction.ledger_id == workspace.id,
            )
        )
    )
    if len(transactions) != len(set(payload.transaction_ids)):
        raise HTTPException(
            status.HTTP_404_NOT_FOUND, "Algun moviment no existeix o no es d'aquest espai"
        )

    merchants: set[int] = set()
    for transaction in transactions:
        transaction.category_id = category.id if category else None
        transaction.category_source = CategorySource.USER
        transaction.category_confidence = 1.0
        transaction.needs_review = False
        if transaction.merchant_id:
            merchants.add(transaction.merchant_id)

    if payload.remember_merchant:
        for merchant_id in merchants:
            merchant = db.get(Merchant, merchant_id)
            if merchant is not None:
                remember_merchant_choice(db, merchant, category.id if category else None)

    db.commit()
    return Message(message=f"{len(transactions)} moviments actualitzats")


def _category_or_400(db: Session, workspace: Ledger, category_id: int | None) -> Category | None:
    """La categoria ha de ser d'aquest espai: no n'hi ha de compartides."""
    if category_id is None:
        return None
    category = db.get(Category, category_id)
    if category is None or category.ledger_id != workspace.id:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "La categoria no es d'aquest espai")
    return category


def _get_in_workspace(db: Session, workspace: Ledger, transaction_id: int) -> Transaction:
    transaction = db.get(Transaction, transaction_id)
    if transaction is None or transaction.ledger_id != workspace.id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Moviment no trobat")
    return transaction


@router.get("/{transaction_id}", response_model=TransactionOut)
def get_transaction(transaction_id: int, db: DbSession, workspace: Workspace):
    return to_out(_get_in_workspace(db, workspace, transaction_id))


@router.patch("/{transaction_id}", response_model=TransactionOut)
def update_transaction(
    transaction_id: int,
    payload: TransactionUpdate,
    db: DbSession,
    user: CurrentUser,
    workspace: EditableWorkspace,
):
    """Actualitza un moviment. Canviar la categoria es una decisio de l'usuari
    i, per defecte, s'aprofita per recordar-la per a tot el comerc de l'espai."""
    transaction = _get_in_workspace(db, workspace, transaction_id)
    data = payload.model_dump(exclude_unset=True)
    remember = data.pop("remember_merchant", True)
    create_rule = data.pop("create_rule", False)

    if "category_id" in data:
        category = _category_or_400(db, workspace, data.pop("category_id"))
        category_id = category.id if category else None
        transaction.category_id = category_id
        transaction.category_source = CategorySource.USER
        transaction.category_confidence = 1.0
        transaction.needs_review = False

        _close_suggestion(db, transaction, category_id)

        if remember and transaction.merchant_id:
            merchant = db.get(Merchant, transaction.merchant_id)
            if merchant is not None:
                remember_merchant_choice(db, merchant, category_id)
        if create_rule:
            build_learned_rule(db, transaction, category_id, created_by_id=user.id)

    if "display_description" in data:
        _apply_display_description(db, workspace.id, transaction, data.pop("display_description"))

    for field, value in data.items():
        setattr(transaction, field, value)
    db.commit()
    db.refresh(transaction)
    return to_out(transaction)


def _apply_display_description(
    db: Session, ledger_id: int, transaction: Transaction, alias: str | None
) -> None:
    """Desa l'àlies i, si es un traspàs, el copia a l'altra cama."""
    value = alias.strip()[:200] if alias and alias.strip() else None
    transaction.display_description = value
    if not transaction.transfer_group_id:
        return
    siblings = db.scalars(
        select(Transaction).where(
            Transaction.transfer_group_id == transaction.transfer_group_id,
            Transaction.ledger_id == ledger_id,
            Transaction.id != transaction.id,
        )
    )
    for sibling in siblings:
        sibling.display_description = value


def _close_suggestion(db: Session, transaction: Transaction, category_id: int | None) -> None:
    """Marca si el suggeriment del model local era encertat."""
    if not transaction.merchant_id:
        return
    suggestion = db.scalar(
        select(LlmSuggestion)
        .where(LlmSuggestion.merchant_id == transaction.merchant_id)
        .order_by(LlmSuggestion.created_at.desc())
        .limit(1)
    )
    if suggestion is None or suggestion.accepted is not None:
        return
    suggestion.accepted = suggestion.suggested_category_id == category_id
    suggestion.reviewed_at = utcnow()


@router.get("/{transaction_id}/related", response_model=list[TransactionOut])
def related_transactions(transaction_id: int, db: DbSession, workspace: Workspace, limit: int = 20):
    """Altres moviments del mateix comerc, per veure el patro de despesa."""
    transaction = _get_in_workspace(db, workspace, transaction_id)
    if not transaction.merchant_id:
        return []
    rows = db.scalars(
        select(Transaction)
        .where(
            Transaction.merchant_id == transaction.merchant_id,
            Transaction.ledger_id == workspace.id,
            Transaction.id != transaction.id,
        )
        .order_by(Transaction.booking_date.desc())
        .limit(limit)
    ).all()
    return [to_out(item) for item in rows]

File: src/services/transactions.ts
/**
 * Moviments: consulta, vista i emmascarament.
 *
 * **L'emmascarament es una funcio de privadesa i s'aplica aqui, no a la
 * plantilla.** Quan un moviment te `display_description`, aquell text
 * substitueix el concepte del banc, i el comerç i la contrapart no es mostren
 * ni es poden cercar.
 *
 * En una arquitectura de fragments aixo es un risc real: qualsevol plantilla
 * nova que dibuixes una fila crua se'l saltaria sense que ningu se n'adones.
 * Per aixo tot passa per `vistaMoviment()` i **de `routes/` no s'importa mai
 * el tipus de la fila sencera**.
 */

import {
  and,
  count,
  desc,
  eq,
  gte,
  ilike,
  inArray,
  isNotNull,
  isNull,
  lte,
  not,
  or,
  sql,
  sum,
  type SQL,
} from "drizzle-orm";

import { parseDescription, type OperationType } from "./concepte.ts";
import { hasTag } from "./tags.ts";

import { db } from "../db/client.ts";
import {
  accounts,
  categories,
  merchants,
  recurringOccurrences,
  recurringSeries,
  transactions,
  type CategorySource,
  type TransactionStatus,
} from "../db/schema/index.ts";
import { NotFoundError } from "../lib/http.ts";
import type { MoneyString } from "../lib/money.ts";

/**
 * Un moviment tal com es pot ensenyar.
 *
 * No hi ha ni `raw`, ni `dedupKey`, ni `entryReference`, ni el concepte del
 * banc quan esta emmascarat. Es l'unic tipus que les plantilles accepten.
 */
export interface TransactionView {
  id: number;
  accountId: number;
  accountName: string | null;
  bookingDate: string;
  valueDate: string | null;
  amount: MoneyString;
  currency: string;
  status: TransactionStatus;
  /**
   * El text que es pot ensenyar: l'alias si n'hi ha; si no, el concepte del
   * banc ja parsejat (sense targeta ni comissio).
   */
  description: string;
  /**
   * Text bancari sense PAN/targeta/comissio, per al `title` del boto.
   * Null quan hi ha alias (la dada del banc no s'ensenya).
   */
  descriptionHint: string | null;
  /** Darrers 4 digits de la targeta, o null. Mai amb alias. */
  darrers4: string | null;
  /**
   * Tipus d'operacio deduit del concepte. Null quan hi ha alias (no ensenyem
   * metadades del banc).
   */
  operationType: OperationType | null;
  counterparty: string;
  merchantId: number | null;
  merchantName: string | null;
  categoryId: number | null;
  categoryName: string | null;
  categorySource: CategorySource;
  categoryConfidence: number | null;
  needsReview: boolean;
  transferGroupId: string | null;
  notes: string;
  tags: string[];
  isExcluded: boolean;
  /** Cert si algu n'ha amagat el concepte del banc. */
  isMasked: boolean;
  /** Serie recurrent enllaçada via `recurring_occurrences`, si n'hi ha. */
  seriesId: number | null;
  seriesLabel: string | null;
}

/**
 * Columnes explicites. Mai `select()` a seques sobre `transactions`: la fila
 * sencera duu `raw`, que es la resposta del banc amb noms i IBAN.
 */
const Fields = {
  id: transactions.id,
  accountId: transactions.accountId,
  accountName: accounts.name,
  bookingDate: transactions.bookingDate,
  valueDate: transactions.valueDate,
  amount: transactions.amount,
  currency: transactions.currency,
  status: transactions.status,
  description: transactions.description,
  displayDescription: transactions.displayDescription,
  normalizedDescription: transactions.normalizedDescription,
  counterparty: transactions.counterparty,
  merchantId: transactions.merchantId,
  merchantName: merchants.displayName,
  categoryId: transactions.categoryId,
  categoryName: categories.name,
  categorySource: transactions.categorySource,
  categoryConfidence: transactions.categoryConfidence,
  needsReview: transactions.needsReview,
  transferGroupId: transactions.transferGroupId,
  notes: transactions.notes,
  tags: transactions.tags,
  isExcluded: transactions.isExcluded,
  seriesId: recurringSeries.id,
  seriesLabel: recurringSeries.label,
} as const;

/**
 * La fila tal com surt de la consulta. Les columnes que venen d'un `left
 * join` poden ser nul·les, de manera que s'escriu a ma en lloc de deduir-la
 * de `CAMPS`: deduir-la amagaria justament aquesta nul·litat.
 */
interface RawRow {
  id: number;
  accountId: number;
  accountName: string | null;
  bookingDate: string;
  valueDate: string | null;
  amount: string;
  currency: string;
  status: TransactionStatus;
  description: string;
  displayDescription: string | null;
  normalizedDescription: string;
  counterparty: string;
  merchantId: number | null;
  merchantName: string | null;
  categoryId: number | null;
  categoryName: string | null;
  categorySource: CategorySource;
  categoryConfidence: number | null;
  needsReview: boolean;
  transferGroupId: string | null;
  notes: string;
  tags: string[];
  isExcluded: boolean;
  seriesId: number | null;
  seriesLabel: string | null;
}

/**
 * Converteix una fila en el que es pot ensenyar, aplicant l'emmascarament.
 *
 * **Es l'unica porta.** Si un moviment esta emmascarat, aqui es on el
 * concepte del banc, la contrapart i el comerç desapareixen. Si no, el
 * concepte es parseja nomes per mostrar (sense tocar la BD).
 */
export function transactionView(row: RawRow): TransactionView {
  const emmascarat = row.displayDescription !== null && row.displayDescription !== "";

  if (emmascarat) {
    return {
      id: row.id,
      accountId: row.accountId,
      accountName: row.accountName,
      bookingDate: row.bookingDate,
      valueDate: row.valueDate,
      amount: row.amount,
      currency: row.currency,
      status: row.status,
      description: row.displayDescription ?? "",
      descriptionHint: null,
      darrers4: null,
      operationType: null,
      counterparty: "",
      merchantId: row.merchantId,
      merchantName: null,
      categoryId: row.categoryId,
      categoryName: row.categoryName,
      categorySource: row.categorySource,
      categoryConfidence: row.categoryConfidence,
      needsReview: row.needsReview,
      transferGroupId: row.transferGroupId,
      notes: row.notes,
      tags: row.tags,
      isExcluded: row.isExcluded,
      isMasked: true,
      seriesId: row.seriesId,
      seriesLabel: row.seriesLabel,
    };
  }

  const parsed = parseDescription(row.description);
  const hint = parsed.cleanedOriginal !== parsed.title ? parsed.cleanedOriginal : null;

  return {
    id: row.id,
    accountId: row.accountId,
    accountName: row.accountName,
    bookingDate: row.bookingDate,
    valueDate: row.valueDate,
    amount: row.amount,
    currency: row.currency,
    status: row.status,
    description: parsed.title,
    descriptionHint: hint,
    darrers4: parsed.darrers4,
    operationType: parsed.type,
    counterparty: row.counterparty,
    merchantId: row.merchantId,
    merchantName: row.merchantName,
    categoryId: row.categoryId,
    categoryName: row.categoryName,
    categorySource: row.categorySource,
    categoryConfidence: row.categoryConfidence,
    needsReview: row.needsReview,
    transferGroupId: row.transferGroupId,
    notes: row.notes,
    tags: row.tags,
    isExcluded: row.isExcluded,
    isMasked: false,
    seriesId: row.seriesId,
    seriesLabel: row.seriesLabel,
  };
}

export interface TransactionsFilters {
  accountId: number | null;
  dateFrom: string | null;
  dateTo: string | null;
  categoryIds: number[];
  merchantId: number | null;
  search: string;
  /** Filtre per etiqueta (insensible a majuscules). Null = sense filtre. */
  tag: string | null;
  /** Tipus d'operacio (OR). Buit = tots. */
  operationType: OperationType[];
  /** Darrers 4 digits de targeta (OR). Buit = totes. */
  cards: string[];
  onlyReview: boolean;
  onlyUnclassified: boolean;
  includeTransfers: boolean;
  limit: number;
  offset: number;
}

/** Predicat SQL alineat amb `detectaTipusOperacio` (sobre el concepte cru). */
function typePredicate(type: OperationType): SQL {
  const description = transactions.description;
  switch (type) {
    case "targeta":
      return sql`(
        ${description} ~* '^(COMPRA|PAGO[[:space:]]+(MOVIL|CON[[:space:]]+MOVIL|TARJETA|EN)[[:space:]])'
        OR ${description} ~* '\\yTARJ'
      )`;
    case "transferencia":
      return sql`(
        ${description} ILIKE 'TRANSFERENCIA%'
        OR ${description} ILIKE 'TRANSF %'
        OR ${description} ILIKE 'TRANSF.%'
      )`;
    case "bizum":
      return sql`(
        ${description} ILIKE 'BIZUM%'
        OR ${description} ILIKE 'ENVIO BIZUM%'
      )`;
    case "rebut":
      return sql`(
        ${description} ILIKE 'RECIBO%'
        OR ${description} ILIKE 'ADEUDO%'
      )`;
    case "altres": {
      // `or()` es tipa com a opcional perque accepta zero arguments; aqui n'hi
      // van quatre de fixos, aixi que no pot ser indefinit. Es comprova en
      // lloc d'afirmar-ho amb un `!`.
      const algun = or(
        typePredicate("targeta"),
        typePredicate("transferencia"),
        typePredicate("bizum"),
        typePredicate("rebut"),
      );
      if (algun === undefined) throw new Error("predicatTipus: cap predicat");
      return not(algun);
    }
  }
}

function typeClause(type: OperationType[]): SQL | undefined {
  if (type.length === 0) return undefined;
  // Si hi ha tots els tipus, no cal filtrar.
  if (type.length === 5) return undefined;
  return or(...type.map(typePredicate));
}

/** El concepte conte aquests 4 digits com a bloc (no enganxats a mes digits). */
function cardPredicate(v: string): SQL {
  return sql`${transactions.description} ~ ('(^|[^0-9])' || ${v} || '($|[^0-9])')`;
}

function cardClause(cards: string[]): SQL | undefined {
  if (cards.length === 0) return undefined;
  return and(typePredicate("targeta"), or(...cards.map(cardPredicate)));
}

/**
 * Darrers 4 digits de cada targeta feta servir en un compte (o tot el
 * ledger si no se'n dona cap). Es dedueix del concepte, igual que
 * `darrers4` a `vistaMoviment()`: no hi ha cap columna a la BD.
 */
export async function cardsAvailable(
  ledgerId: number,
  accountId: number | null,
): Promise<string[]> {
  const on = and(
    eq(transactions.ledgerId, ledgerId),
    accountId !== null ? eq(transactions.accountId, accountId) : undefined,
    typePredicate("targeta"),
    // Un moviment emmascarat no es pot cercar pel concepte bancari
    // (vistaMoviment): tampoc ha de revelar-hi la targeta.
    or(isNull(transactions.displayDescription), eq(transactions.displayDescription, "")),
  );
  const rows = await db
    .selectDistinct({ description: transactions.description })
    .from(transactions)
    .where(on);

  const trobades = new Set<string>();
  for (const f of rows) {
    const { darrers4 } = parseDescription(f.description);
    if (darrers4) trobades.add(darrers4);
  }
  return [...trobades].toSorted();
}

/**
 * Cerca sobre el text **visible**.
 *
 * Un moviment emmascarat no es pot trobar pel concepte del banc ni per la
 * contrapart: nomes per l'alias que hi ha posat una persona i per les notes.
 * Si no fos aixi, es podria endevinar el que s'ha amagat provant paraules.
 */
function searchClause(patro: string): SQL | undefined {
  return or(
    and(
      isNotNull(transactions.displayDescription),
      or(ilike(transactions.displayDescription, patro), ilike(transactions.notes, patro)),
    ),
    and(
      isNull(transactions.displayDescription),
      or(
        ilike(transactions.description, patro),
        ilike(transactions.normalizedDescription, patro),
        ilike(transactions.counterparty, patro),
        ilike(transactions.notes, patro),
      ),
    ),
  );
}

function condicions(ledgerId: number, f: TransactionsFilters): SQL | undefined {
  const parts: (SQL | undefined)[] = [eq(transactions.ledgerId, ledgerId)];

  if (f.accountId !== null) parts.push(eq(transactions.accountId, f.accountId));
  if (f.dateFrom) parts.push(gte(transactions.bookingDate, f.dateFrom));
  if (f.dateTo) parts.push(lte(transactions.bookingDate, f.dateTo));
  if (f.categoryIds.length > 0) parts.push(inArray(transactions.categoryId, f.categoryIds));
  if (f.merchantId !== null) parts.push(eq(transactions.merchantId, f.merchantId));
  if (f.search.trim()) parts.push(searchClause(`%${f.search.trim()}%`));
  if (f.tag) parts.push(hasTag(f.tag));
  parts.push(typeClause(f.operationType));
  parts.push(cardClause(f.cards));
  if (f.onlyReview) parts.push(eq(transactions.needsReview, true));
  if (f.onlyUnclassified) parts.push(isNull(transactions.categoryId));
  // Els traspassos entre comptes propis no son ni ingres ni despesa: per
  // defecte no surten.
  if (!f.includeTransfers) parts.push(isNull(transactions.transferGroupId));

  return and(...parts);
}

export interface TransactionsPage {
  items: TransactionView[];
  total: number;
  /** Suma dels moviments que encaixen amb els filtres, no nomes de la pagina. */
  totalAmount: MoneyString;
  limit: number;
  offset: number;
}

export async function listTransactions(
  ledgerId: number,
  filters: TransactionsFilters,
): Promise<TransactionsPage> {
  const on = condicions(ledgerId, filters);

  const [summary] = await db
    .select({ n: count(), total: sum(transactions.amount) })
    .from(transactions)
    .where(on);

  const rows = await db
    .select(Fields)
    .from(transactions)
    .leftJoin(accounts, eq(accounts.id, transactions.accountId))
    .leftJoin(merchants, eq(merchants.id, transactions.merchantId))
    .leftJoin(categories, eq(categories.id, transactions.categoryId))
    .leftJoin(recurringOccurrences, eq(recurringOccurrences.transactionId, transactions.id))
    .leftJoin(recurringSeries, eq(recurringSeries.id, recurringOccurrences.seriesId))
    .where(on)
    .orderBy(desc(transactions.bookingDate), desc(transactions.id))
    .limit(filters.limit)
    .offset(filters.offset);

  return {
    items: rows.map(transactionView),
    total: summary?.n ?? 0,
    totalAmount: summary?.total ?? "0.00",
    limit: filters.limit,
    offset: filters.offset,
  };
}

/** Un moviment d'aquest espai, ja llest per ensenyar, o 404. */
export async function transactionInWorkspace(
  id: number,
  ledgerId: number,
): Promise<TransactionView> {
  const [row] = await db
    .select(Fields)
    .from(transactions)
    .leftJoin(accounts, eq(accounts.id, transactions.accountId))
    .leftJoin(merchants, eq(merchants.id, transactions.merchantId))
    .leftJoin(categories, eq(categories.id, transactions.categoryId))
    .leftJoin(recurringOccurrences, eq(recurringOccurrences.transactionId, transactions.id))
    .leftJoin(recurringSeries, eq(recurringSeries.id, recurringOccurrences.seriesId))
    .where(and(eq(transactions.id, id), eq(transactions.ledgerId, ledgerId)))
    .limit(1);

  if (!row) throw new NotFoundError("Aquest moviment no existeix");
  return transactionView(row);
}

/** La fila crua, nomes per als serveis. No arriba mai a cap plantilla. */
export async function transactionRow(id: number, ledgerId: number) {
  const [row] = await db
    .select({
      id: transactions.id,
      ledgerId: transactions.ledgerId,
      merchantId: transactions.merchantId,
      categoryId: transactions.categoryId,
      categorySource: transactions.categorySource,
      normalizedDescription: transactions.normalizedDescription,
      counterparty: transactions.counterparty,
      transferGroupId: transactions.transferGroupId,
      displayDescription: transactions.displayDescription,
    })
    .from(transactions)
    .where(and(eq(transactions.id, id), eq(transactions.ledgerId, ledgerId)))
    .limit(1);

  if (!row) throw new NotFoundError("Aquest moviment no existeix");
  return row;
}

// --- Safata de revisio -------------------------------------------------------

/** Un moviment per revisar, amb la proposta del model local si n'hi ha. */
export interface ReviewItem {
  transaction: TransactionView;
  suggestedCategoryId: number | null;
  suggestedCategoryName: string | null;
  confidence: number | null;
  rationale: string;
}

/**
 * La cua de revisio.
 *
 * El model local **no confirma res pel seu compte**: quan proposa una
 * categoria, el moviment queda marcat per revisar amb la seva confiança i la
 * seva justificacio, i qui decideix es una persona.
 */
export async function reviewQueue(
  ledgerId: number,
  limit = 50,
  offset = 0,
): Promise<{ items: ReviewItem[]; total: number }> {
  const on = and(eq(transactions.ledgerId, ledgerId), eq(transactions.needsReview, true));

  const [summary] = await db.select({ n: count() }).from(transactions).where(on);

  const rows = await db
    .select(Fields)
    .from(transactions)
    .leftJoin(accounts, eq(accounts.id, transactions.accountId))
    .leftJoin(merchants, eq(merchants.id, transactions.merchantId))
    .leftJoin(categories, eq(categories.id, transactions.categoryId))
    .leftJoin(recurringOccurrences, eq(recurringOccurrences.transactionId, transactions.id))
    .leftJoin(recurringSeries, eq(recurringSeries.id, recurringOccurrences.seriesId))
    .where(on)
    .orderBy(desc(transactions.bookingDate), desc(transactions.id))
    .limit(limit)
    .offset(offset);

  const merchantIds = [
    ...new Set(rows.map((f) => f.merchantId).filter((x): x is number => x !== null)),
  ];

  // La proposta mes recent de cada comerç.
  const proposals = new Map<
    number,
    {
      categoryId: number | null;
      categoryName: string | null;
      confidence: number | null;
      rationale: string;
    }
  >();
  if (merchantIds.length > 0) {
    const { llmSuggestions } = await import("../db/schema/index.ts");
    const suggeriments = await db
      .select({
        merchantId: llmSuggestions.merchantId,
        categoryId: llmSuggestions.suggestedCategoryId,
        categoryName: categories.name,
        confidence: llmSuggestions.confidence,
        rationale: llmSuggestions.rationale,
      })
      .from(llmSuggestions)
      .leftJoin(categories, eq(categories.id, llmSuggestions.suggestedCategoryId))
      .where(inArray(llmSuggestions.merchantId, merchantIds))
      .orderBy(llmSuggestions.createdAt);

    for (const s of suggeriments) {
      if (s.merchantId === null) continue;
      proposals.set(s.merchantId, {
        categoryId: s.categoryId,
        categoryName: s.categoryName,
        confidence: s.confidence,
        rationale: s.rationale,
      });
    }
  }

  const items = rows.map((row) => {
    const transaction = transactionView(row);
    // Si el moviment esta emmascarat, la proposta tambe s'amaga: parla del
    // comerç, que es justament el que no s'ha de veure.
    const proposal = transaction.isMasked
      ? undefined
      : row.merchantId !== null
        ? proposals.get(row.merchantId)
        : undefined;

    return {
      transaction,
      suggestedCategoryId: proposal?.categoryId ?? null,
      suggestedCategoryName: proposal?.categoryName ?? null,
      confidence: proposal?.confidence ?? null,
      rationale: proposal?.rationale ?? "",
    };
  });

  return { items, total: summary?.n ?? 0 };
}
